File: khetan_hr/khetan_hr/page/employee_status_repo/employee_status_repo.py
import frappe
from frappe.utils.data import today

@frappe.whitelist()
def employee_data(employee=None, from_date=None, to_date=None, employee_type=None, status=None):
    sql_query = """
    SELECT 
    employee as ID,
    IFNULL(custom_office_or_plant, '') as office_or_plant,
    status as status,
    employee_name as full_name,
    employee_type as employee_type,
    COALESCE(NULLIF(department, ''), '') as department,
    COALESCE(NULLIF(plant, ''), '') as plant,
    DATE_FORMAT(date_of_joining, '%d/%m/%Y') as date_of_joining,
    COALESCE(DATE_FORMAT(relieving_date, '%d/%m/%Y'), '') as relieving_date 
    FROM `tabEmployee`
    WHERE 1=1
    """
    
    if employee:
        sql_query += f" AND employee = '{employee}'"
        
    if from_date and not to_date:
        sql_query += f" AND (date_of_joining >= '{from_date}' OR relieving_date >= '{from_date}')"
    elif to_date and not from_date: 
        sql_query += f" AND (date_of_joining <= '{to_date}' OR relieving_date <= '{to_date}')"
    elif from_date and to_date:
        sql_query += f" AND ((date_of_joining >= '{from_date}' AND date_of_joining <= '{to_date}') OR (relieving_date >= '{from_date}' AND relieving_date <= '{to_date}'))"
        
    if employee_type:
        sql_query += f" AND employee_type = '{employee_type}'"
    if status:
        sql_query += f" AND status='{status}'"        
        
    data = frappe.db.sql(sql_query, as_dict=True)
    
    employee_photos = get_employee_profile_photos()

    for employee_record in data:
        for photo in employee_photos:
            if employee_record["full_name"] == photo["employee_name"]:
                employee_record["photo_url"] = photo["photo_url"]
                break
        else:
            employee_record["photo_url"] = None
    print('\n\n\n',data,'\n\n\n')
    return data

   


def get_employee_profile_photos():
    employees = frappe.get_all("Employee", fields=["name", "employee_name"])
    
    employee_photos = []

   
    for employee in employees:
        photo_attachment = frappe.get_all("File", filters={"attached_to_name": employee.name, "attached_to_doctype": "Employee"}, fields=["file_url"], limit=1)

        if photo_attachment:
            path = photo_attachment[0].file_url
            employee_photos.append({"employee_name": employee.employee_name, "photo_url": path})
        else:
            employee_photos.append({"employee_name": employee.employee_name, "photo_url": None})

    return employee_photos

File: khetan_hr/khetan_hr/page/employee_status_repo/employee_status_repo.js
frappe.pages['employee_status_repo'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Employee Status Report',
		single_column: true
	});

	var employee = ''
	var from_date = ''
	var to_date = ''
	var employee_type = ''
	var status = ''

	let e_field = page.add_field({
        label: 'Employee',
        fieldtype: 'Link',
        fieldname: 'employee',
        options: 'Employee',
        change() {
            employee = e_field.get_value();
            check_filters();
        }
    });

	let f_date = page.add_field({
        label: 'From Date',
        fieldtype: 'Date',
        fieldname: 'from_date',
        change() {
            from_date = f_date.get_value()
            check_filters()
        }
    });

	let t_date = page.add_field({
        label: 'To Date',
        fieldtype: 'Date',
        fieldname: 'to_date',
        change() {
            to_date = t_date.get_value()
            check_filters()
        }
    });
    
	let e_type = page.add_field({
        label: 'Employee Type',
        fieldtype: 'Link',
        fieldname: 'to_date',
		options : 'Employee Type',
        change() {
            employee_type = e_type.get_value()
            check_filters()
        }
    });

	let status_field = page.add_field({
        label: 'Status',
        fieldtype: 'Select',
        fieldname: 'status',
        options: '\nActive\nInactive\nSuspended\nLeft',
        change() {
            status = status_field.get_value();
            check_filters();
        }
    });

    page.set_primary_action('Download', function() {
               downloadEmployeeData();
    }, 'octicon octicon-cloud-download');

    function downloadEmployeeData() {
        frappe.call({
          method: 'khetan_hr.khetan_hr.page.employee_status_repo.employee_status_repo.employee_data',
          args: {
            employee: employee,
            from_date: from_date,
            to_date: to_date,
            employee_type: employee_type,
            status: status
          },
          callback: function(response) {
            var empdata = response.message;
			console.log(empdata)
            generateExcelFile(empdata);
          }
        });
      }
      function generateExcelFile(data) {
        const wb = XLSX.utils.book_new();
        const customHeaders = ['ID', 'Full Name', 'Office or Plant','Employee Type', 'Department', 'Plant', 'Status', 'Relieving Date'];
        const ws = XLSX.utils.aoa_to_sheet([customHeaders, ...data.map(item => {
          const { photo_url, date_of_joining, ...rest } = item;
          return [rest.ID, rest.full_name, rest.office_or_plant, rest.employee_type, rest.department, rest.plant, rest.status, rest.relieving_date];
        })]);
        XLSX.utils.book_append_sheet(wb, ws, 'EmployeeData');
        XLSX.writeFile(wb, 'employee_data.xlsx');
      }
  
	function check_filters(){
        if(employee !== null && from_date !== null && to_date !== null && employee_type !==null && status !==null ){
            get_data(employee, from_date, to_date, employee_type, status);
        } else {
            get_data(null, null, null ,null , null); 
        }
    }

	get_data(null, null, null, null, null);
    
	function get_data(employee, from_date, to_date, employee_type, status){
        frappe.call({
            method: 'khetan_hr.khetan_hr.page.employee_status_repo.employee_status_repo.employee_data',
            args: {
                employee: employee,
                from_date: from_date,
                to_date: to_date,
				employee_type : employee_type,
				status : status
            },
            callback: function(response) {
                var empdata = response.message
                $("#1").remove();
                $(frappe.render_template("employee_status_repo", {empdata: empdata})).appendTo(page.body);
                console.log(response.message)

                
            }
        })
    }  
}